crates/valgrind/src/lib.rs

```diff
@@ -0,0 +1,417 @@
+            return Ok(());
+        }
+        if !(self.is_in_bounds_stack(addr, len) || self.is_in_bounds_heap(addr, len)) {
+            return Err(AccessError::OutOfBounds {
+                addr: addr,
+                len: len,
+            });
+        }
+        for i in addr..addr + len {
+            match self.metadata[i] {
+                MemState::Unallocated => {
+                    return Err(AccessError::InvalidRead {
+                        addr: addr,
+                        len: len,
+                    });
+                }
+                MemState::ValidToWrite => {
+                    return Err(AccessError::InvalidRead {
+                        addr: addr,
+                        len: len,
+                    });
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    pub fn write(&mut self, addr: usize, len: usize) -> Result<(), AccessError> {
+        if !self.flag {
+            return Ok(());
+        }
+        if !(self.is_in_bounds_stack(addr, len) || self.is_in_bounds_heap(addr, len)) {
+            return Err(AccessError::OutOfBounds {
+                addr: addr,
+                len: len,
+            });
+        }
+        for i in addr..addr + len {
+            if let MemState::Unallocated = self.metadata[i] {
+                return Err(AccessError::InvalidWrite {
+                    addr: addr,
+                    len: len,
+                });
+            }
+        }
+        for i in addr..addr + len {
+            self.metadata[i] = MemState::ValidToReadWrite;
+        }
+        Ok(())
+    }
+
+    pub fn free(&mut self, addr: usize) -> Result<(), AccessError> {
+        if !self.mallocs.contains_key(&addr) {
+            return Err(AccessError::InvalidFree { addr: addr });
+        }
+        let len = self.mallocs[&addr];
+        for i in addr..addr + len {
+            if let MemState::Unallocated = self.metadata[i] {
+                return Err(AccessError::InvalidFree { addr: addr });
+            }
+        }
+        self.mallocs.remove(&addr);
+        for i in addr..addr + len {
+            self.metadata[i] = MemState::Unallocated;
+        }
+        Ok(())
+    }
+
+    fn is_in_bounds_heap(&self, addr: usize, len: usize) -> bool {
+        self.max_stack_size <= addr && addr + len <= self.metadata.len()
+    }
+
+    fn is_in_bounds_stack(&self, addr: usize, len: usize) -> bool {
+        self.stack_pointer <= addr && addr + len < self.max_stack_size
+    }
+
+    pub fn update_stack_pointer(&mut self, new_sp: usize) -> Result<(), AccessError> {
+        if new_sp > self.max_stack_size {
+            return Err(AccessError::OutOfBounds {
+                addr: self.stack_pointer,
+                len: new_sp - self.stack_pointer,
+            });
+        } else if new_sp < self.stack_pointer {
+            for i in new_sp..self.stack_pointer + 1 {
+                // +1 to account for sp == max_stack_size (?)
```

Comment:
I'm somewhat surprised this is needed -- could we document the case where it is if so (example addresses or whatnot)?